QHyper/solvers/QAOA/Pennylane.py

```diff
@@ -0,0 +1,236 @@
+    backend : str
+        name of the backend (default "default.qubit")
+    dev : qml.device
+        PennyLane quantum device
+    """
+
+    def __init__(self, **kwargs) -> None:
+        """
+        Parameters
+        ----------
+        problem : Problem
+            object of class Problem, defining problem
+        angles: list[float]
+            angles for QAOA, also initial values for optimizers
+        optimizer : Optmizer
+            optimizer that will be used to optimize angles (default None)
+        layers : int
+            number of layers in circuit (default 3)
+        mixer : str
+            mixer name (currently only "X" is supported) (default "X")
+        weights : list[float]
+            needed for converting Problem to QUBO, if QUBU is already provided
+            weights should be [1] (default None)
+        hyperoptimizer : HyperparametersOptimizer
+            optimizer to tune the weights (default None)
+        backend : str
+            name of the backend (default "default.qubit")
+        """
+
+        self.problem: Problem = kwargs.get("problem")
+        self.angles: list[float] = kwargs.get("angles")
+
+        self.optimizer: Optimizer = kwargs.get("optimizer", None)
+        self.layers: int = kwargs.get("layers", 3)
+        self.mixer: str = kwargs.get("mixer", "X")
+        self.weights: list[float] = kwargs.get("weights", None)
+        self.hyperoptimizer: HyperparametersOptimizer = kwargs.get("hyperoptimizer", None)
+        self.backend: str = kwargs.get("backend", "default.qubit")
+
+        self.dev = qml.device(self.backend, wires=self.problem.wires)
+
+    def _create_cost_operator(self, weights) -> qml.Hamiltonian:
+        ingredients = [self.problem.objective_function] + self.problem.constraints
+        cost_operator = ""
+        if len(weights) != len(ingredients):
+            raise Exception(
+                f"Number of provided weights ({len(weights)}) is different from number of ingredients ({len(ingredients)})")
+        for weight, ingredient in zip(weights, ingredients):
+            cost_operator += f"+{weight}*({ingredient})"
+        return parse_hamiltonian(cost_operator)
+
+    def _hadamard_layer(self):
+        for i in range(self.problem.wires):
+            qml.Hadamard(i)
+    
+    def _create_mixing_hamitonian(self) -> qml.Hamiltonian:
+        hamiltonian = qml.Hamiltonian([], [])
+        if self.mixer == "X":
+            for i in range(self.problem.wires):
+                hamiltonian += qml.Hamiltonian([1/2], [qml.PauliX(i)])
+        return hamiltonian
+
+    def _circuit(self, params, cost_operator: qml.Hamiltonian):
+        def qaoa_layer(gamma, beta):
+            qml.qaoa.cost_layer(gamma, cost_operator)
+            qml.qaoa.mixer_layer(beta, self._create_mixing_hamitonian()) 
+
+        self._hadamard_layer()
+        qml.layer(qaoa_layer, self.layers, params[0], params[1]) 
+
+    def get_expval_func(self, weights: list[float]):
```

Comment:
Many methods in this file don't have the return type specified. If it's not done on purpose, it should be fixed.

Reply:
I will leave it for now, we will have to run mypy anyway.